athena/kernel/src/arch/x64/context.rs

```diff
@@ -0,0 +1,317 @@
+pub fn init() {
+    sce_enable();
+}
+
+#[derive(Debug, Copy, Clone, Default)]
+#[allow(missing_docs, dead_code)]
+#[repr(packed)]
+struct PreservedRegisters {
+    pub rbx: u64,
+    pub rbp: u64,
+    pub r12: u64,
+    pub r13: u64,
+    pub r14: u64,
+    pub r15: u64,
+}
+
+#[derive(Debug, Copy, Clone, Default)]
+#[allow(missing_docs, dead_code)]
+#[repr(packed)]
+struct ScratchRegisters {
+    pub rax: u64,
+    pub rcx: u64,
+    pub rdx: u64,
+    pub rsi: u64,
+    pub rdi: u64,
+    pub r8: u64,
+    pub r9: u64,
+    pub r10: u64,
+    pub r11: u64,
+}
+
+#[derive(Debug, Copy, Clone, Default)]
+#[repr(packed)]
+#[allow(missing_docs, dead_code)]
+struct Regs {
+    pub preserved: PreservedRegisters,
+    pub scratch: ScratchRegisters,
+
+    pub rsp: u64,
+    pub rip: u64,
+    pub rflags: u64,
+}
+
+/// A generic runnable context.
+#[derive(Debug)]
+#[repr(transparent)]
+pub struct Context(ContextVariant);
+
+#[derive(Debug)]
+enum ContextVariant {
+    /// Kernel thread.
+    KThread(KThread),
+}
+
+/// A kernel thread context.
+#[derive(Debug)]
+struct KThread {
+    regs: Regs,
+    _stack_page: Page<Size4KiB>,
+}
+
+impl Regs {
+    /// Construct a new
+    pub fn new() -> Self {
+        Default::default()
+    }
+
+    /// Performs a context switch, to the `self` state, saving the preserved registers in `previous`.
+    ///
+    /// This has the effect of 1) switching execution context to the saved state in `self` and 2)
+    /// Saving the current state of execution to `current`, such that on a follow up switch, it will
+    /// return back to the caller as if this function had been a no-op.
+    ///
+    /// This function will also set interrupts back on.
+    #[naked]
+    pub unsafe extern "sysv64" fn switch(restore: *const Self, store: *mut Self) {
+        unsafe {
+            asm!(
+                // Save current state
+
+                // Return pointer
+                "pop rax",
+                "mov [rsi + 8*16], rax",
+                "mov [rsi + 8*15], rsp",
+                "mov [rsi], rbx",
+                "mov [rsi + 8], rbp",
+                "mov [rsi + 8*2], r12",
+                "mov [rsi + 8*3], r13",
+                "mov [rsi + 8*4], r14",
+                "mov [rsi + 8*5], r15",
+                // Restore the registers
+                "mov rbx, [rdi]",
+                "mov rbp, [rdi + 8]",
+                "mov r12, [rdi + 8*2]",
+                "mov r13, [rdi + 8*3]",
+                "mov r14, [rdi + 8*4]",
+                "mov r15, [rdi + 8*5]",
+                "mov rax, [rdi + 8*6]",
+                "mov rcx, [rdi + 8*7]",
+                "mov rdx, [rdi + 8*8]",
+                "mov rsi, [rdi + 8*9]",
+                "mov r8, [rdi + 8*11]",
+                "mov r9, [rdi + 8*12]",
+                "mov r10, [rdi + 8*13]",
+                "mov r11, [rdi + 8*14]",
+                "mov rsp, [rdi + 8*15]",
+                // rflags: TODO: Mask interrupts?
```

Comment:
I think it may be okay to not mask interrupts, and actually, it's probably correct to not reenable interrupts necessarily, only if they were enabled in the first place. New kthreads would need the rflags int bit to be initialized. If we get an interrupt somewhere here, all registers will be restored back to this point anyway if I'm not mistaken.

Reply:
Yes and no. It's a bit weird if we are jumping from main. I believe there's a better way to handle it by forgetting about the `current` context and instead having main be a context as everything else.

Reply:
Leaving it to be fixed at a later point.